vpr/src/route/connection_router.cpp

```diff
@@ -685,13 +685,16 @@ void ConnectionRouter<Heap>::evaluate_timing_driven_node_costs(t_heap* to,
     float switch_Tdel = rr_switch_inf_[iswitch].Tdel;
     float switch_Cinternal = rr_switch_inf_[iswitch].Cinternal;
 
+    const auto& device_ctx = g_vpr_ctx.device();
+    const auto& rr_graph = device_ctx.rr_graph;
+
     //To node info
-    auto rc_index = rr_nodes_.node_rc_index(RRNodeId(to_node));
+    auto rc_index = rr_graph.node_rc_index(RRNodeId(to_node));
```

Comment:
The connection_router object actually caches the rr_graph, in order to avoid the use of global variables. I suggest to add the RRGraphView to the cache here: https://github.com/verilog-to-routing/vtr-verilog-to-routing/blob/2e0bafe202e3f8cb4e1f099002f7272f4705068d/vpr/src/route/connection_router.h#L27-L44

Then you just use the cached copy ``rr_graph_`` in these functions.

Reply:
@tangxifan I am trying to add the RRGraphView to the cache with 

`const RRGraphView& rr_graph,` and

`, rr_graph_(rr_graph)`

But when I try this, the compiler gives me the following warning

`error: use of deleted function ‘RRGraphView::RRGraphView(const RRGraphView&)’`

Do you know what I am missing? I know that the copy constructors have been disabled in RRGraphView, but perhaps I am supposed to copy to cache in a different way.

Reply:
Yes. You can consider to use the pointer 

https://github.com/verilog-to-routing/vtr-verilog-to-routing/blob/2e0bafe202e3f8cb4e1f099002f7272f4705068d/vpr/src/route/rr_graph_uxsdcxx_serializer.h#L1853

which works fine in the reader/writer class.

Reply:
Anyhow, I think the use of ``const RRGraphView&`` is better. But not sure the compiler spits out such error. 
@hzeller If you have time, can you look into this and share your opinion? Thanks!

Reply:
This code snippet looks like a call to the deleted copy constructor (but I am just going by the conversation, so I don't see the code in context):
`, rr_graph_(rr_graph)`

Reply:
@vaughnbetz, I think you are correct. I have implemented the `RRGraphView*` as @tangxifan suggested, but we could also modify `RRGraphView.h` so that the copy constructor was no longer deleted if that is what we wanted to do.